### PR TITLE
fixes 404 no longer returned when error with code ENOENT is thrown.

### LINKED ISSUE #11406: Cannot return a 404 from getInitialProps anymore  by throwing an error with code='ENOENT'
# Bug report

## Describe the bug

Until next.js 9.2.1 I could return a 404 for a page when throwing an error with `code = 'ENOENT'` in `getInitialProps`. That no longer works in 9.3.2, now a 500 is returned. This is bad for SEO, we need to be able to throw 404s.

## To Reproduce

1. In getInitialProps, throw an error like this one:
```
export class Custom404Error extends Error {
    code: string

    constructor(message?: string) {
        super(message)
        this.code = 'ENOENT' // triggers a 404 in next.js
    }
}
```

2. Open that page in the browser

## Expected behavior

Opening the page in the browser should return status code 404.

## System information

- OS: macOS Catalina
- Browser (if applies) [e.g. chrome, safari]
- Version of Next.js: 9.3.2

## Additional context

If there is another recommended way to return 404s from getInitialProps, I'd be happy about pointers.
Could the regression (or change of behavior, since `ENOENT` was kind of an internal behavior) be caused by #10572?

### PR DESCRIPTION
**Pull request by fabb**
Restore ENOENT handling during rendering to allow returning 404 statusCodes from getInitialProps.

Fixes #11406

This PR reverts an unwanted code change done in refactoring PR #10476.

Now when throwing an error with `error.code='ENOENT'` in `getInitialProps` correctly returns a 404 statusCode in the browser again.

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| buildDuration | 11.6s | 11.1s | -445ms |
| nodeModulesSize | 52.8 MB | 52.8 MB | ⚠️ +408 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.24 kB | 6.24 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..c6c1.js gzip | 10.1 kB | 10.1 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.2 kB | 56.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.77 kB | 4.77 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.71 kB | 6.71 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| index.html gzip | 917 B | 917 B | ✓ |
| link.html gzip | 925 B | 925 B | ✓ |
| withRouter.html gzip | 915 B | 915 B | ✓ |
| Overall change | 2.76 kB | 2.76 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| buildDuration | 12s | 12.3s | ⚠️ +314ms |
| nodeModulesSize | 52.8 MB | 52.8 MB | ⚠️ +408 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.24 kB | 6.24 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..c6c1.js gzip | 10.1 kB | 10.1 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.2 kB | 56.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.77 kB | 4.77 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.71 kB | 6.71 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | fabb/next.js fix-404-status-code | Change |
| - | - | - | - |
| _error.js gzip | 294 kB | 294 kB | -196 B |
| 404.html gzip | 1.32 kB | 1.32 kB | ✓ |
| hooks.html gzip | 958 B | 958 B | ✓ |
| index.js gzip | 293 kB | 294 kB | ⚠️ +497 B |
| link.js gzip | 301 kB | 302 kB | ⚠️ +376 B |
| routerDirect.js gzip | 301 kB | 300 kB | -178 B |
| withRouter.js gzip | 300 kB | 300 kB | ⚠️ +140 B |
| Overall change | 1.49 MB | 1.49 MB | ⚠️ +639 B |

</details>

</details>

**Review by Timer — CHANGES_REQUESTED** [review]
This behavior needs a test before we can land this. In addition, it should not work for `getServerSideProps`.

Just to be clear: **we do not suggest apps use this**. You are relying on an undocumented and unintentional behavior, and this was never meant to be used like this. It's being added for backwards compatibility because it seems a lot of apps found and started using this.

**Comment by fabb** [comment]
I migrated away from using this hack as I described in the linked issue. I‘ll keep this open still because others might still use this hack and not have the resources for refactoring.

**Review by ijjk — APPROVED** [review]

**Review by timneutkens — APPROVED** [review]

**Review by Timer — CHANGES_REQUESTED** [review]
We need a test to check that this does not work for getServerSideProps. Also, this should warn the user that relying on the behavior is wrong and can change at any time.

**Comment by ijjk** [comment]
Added a warning when an error is thrown with `ENOENT` to trigger the 404 and added test for the warning, also made sure to copy the existing test for this from `getServerSideProps` to the `prerender` suite to ensure it's handled there also

https://github.com/zeit/next.js/blob/13cf97055ff509826be6d21f66424ae9d5c3c21f/test/integration/getserversideprops/test/index.test.js#L214-L225

File: errors/getinitialprops-enoent.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# ENOENT from getInitialProps
+
+#### Why This Error Occurred
+
+In one of your pages you threw an error with the code `ENOENT`, this was previously an internal feature used to trigger rendering the 404 page and should not be relied on.
+
+#### Possible Ways to Fix It
+
+Handle rendering the 404 state on the current page instead of throwing and trying to render a separate page.
```
Comment (timneutkens) [inline errors/getinitialprops-enoent.md:9]: This should reference setting the statuscode in getServerSideProps or getInitialProps.

File: errors/getinitialprops-enoent.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+### Useful Links
+
+- [Google 404 Error Guide](https://developers.google.com/search/docs/guides/fix-search-javascript)
```
Comment (timneutkens) [inline errors/getinitialprops-enoent.md:13]: The 404 docs should be referenced here too I think

Comment (ijjk) [reply errors/getinitialprops-enoent.md:15]: Added

File: errors/getinitialprops-enoent.md (hunk 1 above)
Comment (ijjk) [reply errors/getinitialprops-enoent.md]: Updated to mention the status code can be set in these methods also

**Comment by timneutkens** [comment]
Looking at the issue and this PR it seems that there were only a few people that relied on these internals and adding extra checks is not warranted.

**Pull request closed without merging**